core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -920,6 +923,9 @@ private[spark] class TaskSetManager(
         s" be re-executed (either because the task failed with a shuffle data fetch failure," +
         s" so the previous stage needs to be re-run, or because a different copy of the task" +
         s" has already succeeded).")
+    } else if (sched.stageIdToFinishedPartitions.get(stageId).exists(
+      partitions => partitions.contains(tasks(index).partitionId))) {
+      sched.markPartitionCompletedInAllTaskSets(stageId, tasks(index).partitionId, info)
```

Comment:
`stageIdToFinishedPartitions` is getting updated in the dag scheduler event loop, but here you're querying it outside of the event loop, which is definitely not safe.

(I'm also not really convinced this would solve the problem, but I'm afraid I have to spend a bit more time paging it all back in  ...)

Reply:
Yes that is correct, `stageIdToFinishedPartitions` is getting updated in the event loop and is being queried outside the event loop within the TaskSchedulerImpl lock. However, I did not get as to why is it not safe as the only thing that can happen is that while querying, TSM does not find the partition to be present in the HashSet while it is getting updated, but it will definitely catch this in the next check. I could be wrong though.

Reply:
hashmaps are totally unsafe to be used for multiple threads -- its not just getting inconsistent values, its that the hashmap may be in some undefined state b/c of rehashing.  see eg. http://javabypatel.blogspot.com/2016/01/infinite-loop-in-hashmap.html (I just skimmed this but I think it has the right idea).

Reply:
I see, in that case, what if I turn `stageIdToFinishedPartitions` to a ConcurrentHashMap? That should take care of the safety feature.

Reply:
that will take care of the access-during-rehash, but I'm still not sure the values of this are safe.  Here you're querying it from a task-result-getter thread, but you're also updating it from the dag scheduler event loop, with no lock protecting access.

The other proposal is easier to reason about, because it keeps this structure protected by a lock on TaskSchedulerImpl.